app/connectors/TaiConnector.scala

```diff
@@ -33,20 +36,29 @@ class TaiConnector @Inject() (
   val httpClientV2: HttpClientV2,
   servicesConfig: ServicesConfig,
   httpClientResponse: HttpClientResponse,
-  configDecorator: ConfigDecorator
+  configDecorator: ConfigDecorator,
+  featureFlagService: FeatureFlagService
 ) {
 
-  private lazy val taiUrl = servicesConfig.baseUrl("tai")
-  def taxComponents(nino: Nino, year: Int)(implicit
+  private lazy val taiUrl                              = servicesConfig.baseUrl("tai")
+  def taxComponents[A](nino: Nino, year: Int)(reads: Reads[A])(implicit
     hc: HeaderCarrier,
     ec: ExecutionContext
-  ): EitherT[Future, UpstreamErrorResponse, HttpResponse] = {
-    val url = s"$taiUrl/tai/$nino/tax-account/$year/tax-components"
-    httpClientResponse.read(
-      httpClientV2
-        .get(url"$url")
-        .transform(_.withRequestTimeout(configDecorator.taiTimeoutInMilliseconds.milliseconds))
-        .execute[Either[UpstreamErrorResponse, HttpResponse]](readEitherOf(readRaw), ec)
-    )
-  }
+  ): EitherT[Future, UpstreamErrorResponse, Option[A]] =
+    featureFlagService.getAsEitherT(TaxComponentsToggle).flatMap { toggle =>
+      if (toggle.isEnabled) {
+        val url = s"$taiUrl/tai/$nino/tax-account/$year/tax-components"
+        httpClientResponse
+          .read(
+            httpClientV2
+              .get(url"$url")
+              .transform(_.withRequestTimeout(configDecorator.taiTimeoutInMilliseconds.milliseconds))
+              .execute[Either[UpstreamErrorResponse, HttpResponse]](readEitherOf(readRaw), ec)
+          )
+          .map(result => result.json.asOpt[A](reads))
```

Comment:
don't use asOpt. If it fails to parse the json it does not log anything.
It's better to recover to log the exception and return None

Reply:
Good point. I've applied change now